### DISCUSSION TITLE
extend script for changing creatin date to all children

### DISCUSSION THREAD

**Discussion by Olifair** [freeplane/freeplane#1662] [Help]
Hello, 
I have been using quite regularly a one line script to change the creation date of a node : 
I go "edit script", "new script", copy paste, change the date in the line and then run, exit without saving.
`node.setCreatedAt(Date.parse("yyyy-MM-dd","2023-08-22"))`
Is there an easy way to extend the script to the node and all children  ? 

Thanks in advance

Oliver

**Comment by euu2021** [comment]
I have 2 questions:

Would it be better for you if the script worked like this?
- user assign the script to a hotkey. This needs to be done only once
- when the hotkey is fired, a text box window, or a date picker, pop up, so the user types the desired date
- user press ok and the date is applied to the currently selected node.

> Is there an easy way to extend the script to the node and all children ?

Do you mean: user selects one node and apply the script; all the nodes of the branch have their creation dates changed.

**Reply by Olifair** [reply]
Well, clearly having this functionality as a real groovy script which would open a text box or even a date picker would be luxury :-)
And yes, it would be great if there could be the option that one node is selected and all nodes of the branch have their creation dates changed

**Reply by euu2021** [reply]
This script creates a text box where you can type the date in the format "yyyy-MM-dd". 

The new date is applied to the whole branch. You can change this behavior in the line `node.findAll().each {it.setCreatedAt(newDate)}`.

The text box is pre-filled with a default date ("2024-01-01") to guide the format and provide a starting point. To change the default date, locate the line in the script that says String defaultDate = "2024-01-01" and replace "2024-01-01" with your preferred default date in the same format.

```
import javax.swing.JOptionPane
import java.text.SimpleDateFormat

// Default date to be shown in the text box
String defaultDate = "2024-01-01"

// Display input dialog with default date
String inputDate = JOptionPane.showInputDialog(null, "Enter the date in format yyyy-MM-dd", defaultDate)

// Check if the user has entered a date
if (inputDate != null && !inputDate.isEmpty()) {
    try {
        // Parse the user input to a Date object
        Date newDate = new SimpleDateFormat("yyyy-MM-dd").parse(inputDate)

        // Set the creation date of the current node branch
        node.findAll().each {it.setCreatedAt(newDate)}
    } catch (Exception e) {
        // Show error message if the date format is incorrect
        JOptionPane.showMessageDialog(null, "Invalid date format. Please enter the date in format yyyy-MM-dd.", "Error", JOptionPane.ERROR_MESSAGE)
    }
} else {
    // Show message if user cancels or enters no date
    JOptionPane.showMessageDialog(null, "No date entered.", "Info", JOptionPane.INFORMATION_MESSAGE)
}
```

**Reply by euu2021** [reply]
PS: 
May be relevant to mention: you can assign the script to a hotkey.

**Reply by euu2021** [reply]
Actually, I noticed that the behavior is not ideal for when multiple nodes are selected. And, it will be more nice if the typed date is saved to be the default text the next time the script is used.

I will look at it tomorrow.

**Reply by euu2021** [reply]
I did some changes. Now,
- the date is applied to all selected nodes (and its branch)
- the typed date shows up as default the next time the script is fired. This works per map (each map has its own memory), and is saved if the map is closed.

```
// @ExecutionModes({ON_SINGLE_NODE})

import javax.swing.JOptionPane
import java.text.SimpleDateFormat

def myNodes = [] + c.selecteds

// Default date to be shown in the text box
String defaultDate
if (node.mindMap.storage['lastDateEntered'] == null) {
    defaultDate = "2024-01-01"
}
else {
    defaultDate = node.mindMap.storage['lastDateEntered']
}

// Display input dialog with default date
String inputDate = JOptionPane.showInputDialog(null, "Enter the date in format yyyy-MM-dd", defaultDate)

// Check if the user has entered a date
if (inputDate != null && !inputDate.isEmpty()) {
    try {
        // Parse the user input to a Date object
        Date newDate = new SimpleDateFormat("yyyy-MM-dd").parse(inputDate)

        // Set the creation date of the current node branch
        myNodes.each{
            it.findAll().each {it.setCreatedAt(newDate)}
        }
        node.mindMap.storage['lastDateEntered'] = inputDate
    } catch (Exception e) {
        // Show error message if the date format is incorrect
        JOptionPane.showMessageDialog(null, "Invalid date format. Please enter the date in format yyyy-MM-dd.", "Error", JOptionPane.ERROR_MESSAGE)
    }
} else {
    // Show message if user cancels or enters no date
    JOptionPane.showMessageDialog(null, "No date entered.", "Info", JOptionPane.INFORMATION_MESSAGE)
}
```

When updating from the previous script, it may be necessary to restart Freeplane, because of the change in the execution mode.

**Reply by Olifair** [reply]
Wow, this is really great. And works nicely on all selected nodes and their children. Thanks very much, this will save me so much time, it's clearly much more elegant than copy/pasting a script into each node. 
I just realised something funny : if you type 2023-14-15 the date will be 2024-02-15, so if the month is >12, it will be automatically MM-12 :-)
Cheers

Oliver
